ctpolicy/ctpolicy.go

```diff
@@ -49,13 +53,113 @@ func (group *LogGroupInfo) setMinInclusions(i int) error {
 		}
 	}
 }
 
+// satisfyMinimalInclusion returns whether number of positive weights is
+// bigger or equal to minimal inclusion number.
+func (group *LogGroupInfo) satisfyMinimalInclusion(weights map[string]float32) bool {
+	nonZeroNum := 0
+	for logURL, w := range weights {
+		if group.LogURLs[logURL] && w > 0.0 {
+			nonZeroNum++
+			if nonZeroNum >= group.MinInclusions {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// SetLogWeights applies suggested weights to the Log-group. Does not reset
+// weights and returns error when there are not enough positive weights
+// provided to reach minimal inclusion number.
+func (group *LogGroupInfo) SetLogWeights(weights map[string]float32) error {
+	if !group.satisfyMinimalInclusion(weights) {
+		return fmt.Errorf("trying to assign weights %v resulting into unability to reach minimal inclusion number %d", weights, group.MinInclusions)
+	}
+	group.wMu.Lock()
+	defer group.wMu.Unlock()
+	// All group weights initially reset to 0.0
+	for logURL := range group.LogURLs {
+		group.LogWeights[logURL] = 0.0
+	}
+	for logURL, w := range weights {
+		if group.LogURLs[logURL] {
+			group.LogWeights[logURL] = w
+		}
+	}
+	return nil
+}
+
+// setLogWeight tries setting the weight for a single Log of the Log-group.
+// Does not reset the weight and returns error if weight is non-positive and
+// its setting will result innto unability to reach minimal inclusion number.
+func (group *LogGroupInfo) SetLogWeight(logURL string, w float32) error {
+	if !group.LogURLs[logURL] {
+		return fmt.Errorf("trying to assign weight to Log %q not belonging to the group", logURL)
+	}
+	newWeights := make(map[string]float32)
+	for l, wt := range group.LogWeights {
+		newWeights[l] = wt
+	}
+	newWeights[logURL] = w
+	if !group.satisfyMinimalInclusion(newWeights) {
+		return fmt.Errorf("assigning weight %v to Log %q will result into unability to reach minimal inclusion number %d", w, logURL, group.MinInclusions)
+	}
+	group.wMu.Lock()
+	defer group.wMu.Unlock()
+	group.LogWeights = newWeights
+	return nil
+}
+
+// getSubmissionSession produces list of log-URLs of the Log-group.
+// Order of the list is weighted random defined by Log-weights within the group
+func (group *LogGroupInfo) GetSubmissionSession() []string {
+	if len(group.LogURLs) == 0 {
+		return make([]string, 0)
+	}
+	session := make([]string, 0)
+	// modelling weighted random with exclusion
+
+	var sum float32
+	for _, w := range group.LogWeights {
+		sum += w
+	}
+	processedURLs := make(map[string]bool)
+	for logURL := range group.LogURLs {
+		processedURLs[logURL] = false
+	}
+
+	group.wMu.RLock()
+	defer group.wMu.RUnlock()
+	for i := 0; i < len(group.LogURLs); i++ {
+		if sum <= 0.0 {
+			break
+		}
+		r := rand.Float32() * sum
+		for logURL, w := range group.LogWeights {
+			if processedURLs[logURL] || w <= 0.0 {
+				continue
+			}
+			r -= w
+			if r < 0.0 {
+				session = append(session, logURL)
+				processedURLs[logURL] = true
+				sum -= w
```

Comment:
continue here?

Reply:
Would it be simpler to test if you broke out a single round of "WeightedSample" into a separate function?

Reply:
I'm afraid not. As those are not independent events: we'd like to avoid sending same cert into same Log twice.